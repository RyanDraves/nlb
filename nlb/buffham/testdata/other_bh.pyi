# @generated by Buffham
import dataclasses
from typing import Self

from emb.network.serialize import bh_cobs
from emb.network.transport import transporter
from nlb.buffham import bh

OTHER_CONSTANT = 2


@dataclasses.dataclass
class Ping:
    """This file shares a message name with `sample.bh`
    """

    pong: int

    def serialize(self) -> bytes: ...

    @classmethod
    def deserialize(cls, buffer: bytes) -> tuple[Self, int]: ...

class OtherSerializer(bh_cobs.BhCobs):
    def __init__(self, registry: bh_cobs.Registry | None = None): ...

class OtherNode[Transporter: transporter.TransporterLike](bh.BhNode[OtherSerializer, Transporter]):
    def __init__(self, serializer: OtherSerializer | None = None, transporter: Transporter | None = None): ...

PONG: bh.Transaction[Ping, Ping] = ...
