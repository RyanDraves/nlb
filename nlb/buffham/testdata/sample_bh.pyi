# @generated by Buffham
import dataclasses
import enum
from typing import Self, Type

from emb.network.serialize import bh_cobs
from emb.network.transport import transporter
from nlb.buffham import bh

from nlb.buffham.testdata import other_bh

# This is a constant in the global scope
MY_CONSTANT = 4
CONSTANT_STRING = "Hello, world!"  # constants can have inline comments
# Constants may reference other constants with {brackets}
COMPOSED_CONSTANT = 2 + MY_CONSTANT + other_bh.OTHER_CONSTANT


@dataclasses.dataclass
class Ping:
    """A message comment
    """

    # Add some comments here
    ping: int

    def serialize(self) -> bytes: ...

    @classmethod
    def deserialize(cls, buffer: bytes) -> tuple[Self, int]: ...

@dataclasses.dataclass
class FlashPage:
    """
    A bunch of message comments,
    in a block-like pattern.

    All of these belong to `FlashPage`

    """

    address: int
    # Another field comment
    data: list[int]  # What about some in-line comments for fields?
    # This comment belongs to `read_size`
    read_size: int

    def serialize(self) -> bytes: ...

    @classmethod
    def deserialize(cls, buffer: bytes) -> tuple[Self, int]: ...

@dataclasses.dataclass
class LogMessage:
    message: str

    def serialize(self) -> bytes: ...

    @classmethod
    def deserialize(cls, buffer: bytes) -> tuple[Self, int]: ...

@dataclasses.dataclass
class NestedMessage:
    flag: int
    message: LogMessage
    numbers: list[int]
    pong: Ping
    other_pong: other_bh.Pong

    def serialize(self) -> bytes: ...

    @classmethod
    def deserialize(cls, buffer: bytes) -> tuple[Self, int]: ...

REGISTRY: dict[int, Type[bh.BuffhamLike]] = ...

class SampleSerializer(bh_cobs.BhCobs):
    def __init__(self, registry: bh_cobs.Registry | None = None): ...

class SampleNode[Transporter: transporter.TransporterLike](bh.BhNode[SampleSerializer, Transporter]):
    def __init__(self, serializer: SampleSerializer | None = None, transporter: Transporter | None = None): ...

PING: bh.Transaction[other_bh.Pong, LogMessage] = ...
# Transaction comment
FLASH_PAGE: bh.Transaction[FlashPage, FlashPage] = ...
READ_FLASH_PAGE: bh.Transaction[FlashPage, FlashPage] = ...

class PublishIds(enum.Enum):
    # Publish comment
    LOG_MESSAGE = 4
