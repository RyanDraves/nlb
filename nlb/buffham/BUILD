load("@aspect_rules_py//py:defs.bzl", "py_binary", "py_library")
load("@pip//:requirements.bzl", "requirement")
load("//bzl/macros:python.bzl", "py_test")

py_library(
    name = "bh",
    srcs = ["bh.py"],
    visibility = ["//visibility:public"],
    deps = [
        "//emb/network/node",
        "//emb/network/transport:transporter",
        "//nlb/util:dataclass",
    ],
)

py_binary(
    name = "buffham",
    srcs = ["buffham.py"],
    visibility = ["//visibility:public"],
    deps = [
        ":parser",
        ":py_generator",
        "//nlb/util:click_utils",
        requirement("rich-click"),
    ],
)

py_library(
    name = "engine",
    srcs = ["engine.py"],
    visibility = ["//visibility:public"],
    deps = [
        ":parser",
        "//nlb/util:dataclass",
    ],
)

py_test(
    name = "engine_test",
    srcs = ["engine_test.py"],
    deps = [
        ":engine",
    ],
)

py_library(
    name = "parser",
    srcs = ["parser.py"],
    visibility = ["//visibility:public"],
    deps = [
    ],
)

py_test(
    name = "parser_test",
    srcs = ["parser_test.py"],
    data = glob(["testdata/*.bh"]),
    deps = [
        ":parser",
    ],
)

py_library(
    name = "py_generator",
    srcs = ["py_generator.py"],
    visibility = ["//visibility:public"],
    deps = [
        ":parser",
    ],
)

py_test(
    name = "py_generator_test",
    srcs = ["py_generator_test.py"],
    data = glob([
        "testdata/*",
    ]),
    deps = [
        ":engine",
        ":parser",
        ":py_generator",
        ":bh",
        # Transitive dependency of our generated code
        "//emb/network/serialize:bh_cobs",
    ],
)
