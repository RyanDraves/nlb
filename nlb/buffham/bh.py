"""Utilities for Buffham."""

import dataclasses
import pathlib
from typing import Protocol, Self, Type, cast

from emb.network.node import node
from emb.network.transport import transporter
from nlb.util import dataclass


class BuffhamLike(dataclass.DataclassLike, Protocol):
    """Protocol for Buffham-like objects."""

    def serialize(self) -> bytes: ...

    @classmethod
    def deserialize(cls, buffer: bytes) -> tuple[Self, int]: ...


class BhSerializer(Protocol):
    def serialize(self, msg: BuffhamLike, request_id: int) -> bytes: ...

    def deserialize(self, data: bytes) -> tuple[int, BuffhamLike]: ...


class BhNode[
    Bh_Serializer: BhSerializer,
    CommsTransporter: transporter.TransporterLike,
    LogTransporter: transporter.TransporterLike,
](node.NlbNode[Bh_Serializer, CommsTransporter, LogTransporter]):
    """Generic node that uses buffham datatypes for messages."""

    def _transact(
        self, message: dataclass.DataclassLike, request_id: int
    ) -> dataclass.DataclassLike:
        return super()._transact(message, request_id)


@dataclasses.dataclass
class Transaction[S: BuffhamLike, R: BuffhamLike]:
    """A typed transaction in a buffham node"""

    request_id: int

    def transact(self, node: BhNode, msg: S) -> R:
        return cast(R, node._transact(msg, self.request_id))


def read_file(path: pathlib.Path, schema: Type[BuffhamLike]) -> BuffhamLike:
    """Read a binary Buffham file (.bhb) from a schema"""
    return schema.deserialize(path.read_bytes())[0]


def write_file(path: pathlib.Path, msg: BuffhamLike) -> None:
    """Write a message to a binary Buffham file (.bhb)"""
    path.write_bytes(msg.serialize())
