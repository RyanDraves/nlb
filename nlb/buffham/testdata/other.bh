# Small file to test imports in `sample.bh`

constant uint8_t other_constant = 2;

message Pong {
    uint8_t pong;
}

transaction pong[Pong, Pong];
