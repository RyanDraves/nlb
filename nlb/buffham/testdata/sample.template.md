# Sample template

Here's a neat sample template where we can generate files that reference Buffham constants, like {{ composed_constant }}.
