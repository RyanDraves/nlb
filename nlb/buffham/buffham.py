import enum
import pathlib

import rich_click as click

from nlb.buffham import cpp_generator
from nlb.buffham import parser
from nlb.buffham import py_generator
from nlb.buffham import template_generator
from nlb.util import click_utils


class Languages(enum.Enum):
    PYTHON = enum.auto()
    CPP = enum.auto()
    TEMPLATE = enum.auto()


@click.command()
@click.option(
    '--input',
    '-i',
    type=click.Path(exists=True, dir_okay=False, path_type=pathlib.Path),
    required=True,
    help='Input Buffham file',
)
@click.option(
    '--output',
    '-o',
    type=click.Path(dir_okay=False, path_type=pathlib.Path),
    required=True,
    help='Output file',
)
@click.option(
    '--secondary-output',
    '-s',
    type=click.Path(dir_okay=False, path_type=pathlib.Path),
    help='Secondary output file (e.g. for Python stubs or .cc files)',
)
@click.option(
    '--template-file',
    '-t',
    type=click.Path(exists=True, dir_okay=False, path_type=pathlib.Path),
    help='Template file',
)
@click.option(
    '--dep',
    type=click.Path(exists=True, dir_okay=False, path_type=pathlib.Path),
    multiple=True,
    help='Dependency buffham file (neccessary for now)',
)
@click.option(
    '--language',
    '-l',
    type=click_utils.EnumChoice(Languages),
    required=True,
    help='Output language',
)
def main(
    input: pathlib.Path,
    output: pathlib.Path,
    secondary_output: pathlib.Path | None,
    template_file: pathlib.Path | None,
    dep: list[pathlib.Path],
    language: Languages,
):
    p = parser.Parser()
    ctx = parser.ParseContext({})

    for d in dep:
        p.parse_file(d, ctx)

    buffham = p.parse_file(input, ctx)

    match language:
        case Languages.PYTHON:
            assert secondary_output is not None
            py_generator.generate_python(ctx, buffham.namespace, output, stub=False)
            py_generator.generate_python(
                ctx, buffham.namespace, secondary_output, stub=True
            )
        case Languages.CPP:
            assert secondary_output is not None
            cpp_generator.generate_cpp(ctx, buffham.namespace, output, hpp=True)
            cpp_generator.generate_cpp(
                ctx, buffham.namespace, secondary_output, hpp=False
            )
        case Languages.TEMPLATE:
            assert template_file is not None
            template_generator.generate_template(
                ctx, buffham.namespace, output, template_file
            )
        case _:
            raise ValueError(f'Unsupported language: {language}')

    print(f'Generated {output}')
    if secondary_output is not None:
        print(f'Generated {secondary_output}')


if __name__ == '__main__':
    main()
