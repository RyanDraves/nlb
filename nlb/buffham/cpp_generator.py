import pathlib

from nlb.buffham import parser

T = ' ' * 4  # Indentation

TYPE_MAP = {
    parser.FieldType.UINT8_T: 'uint8_t',
    parser.FieldType.UINT16_T: 'uint16_t',
    parser.FieldType.UINT32_T: 'uint32_t',
    parser.FieldType.UINT64_T: 'uint64_t',
    parser.FieldType.INT8_T: 'int8_t',
    parser.FieldType.INT16_T: 'int16_t',
    parser.FieldType.INT32_T: 'int32_t',
    parser.FieldType.INT64_T: 'int64_t',
    parser.FieldType.FLOAT32: 'float',
    parser.FieldType.FLOAT64: 'double',
    parser.FieldType.STRING: 'std::string',
    parser.FieldType.BYTES: 'std::vector<uint8_t>',
}


def _to_snake_case(name: str) -> str:
    """Convert a title case name to snake case."""
    return name[0].lower() + ''.join(
        f'_{c.lower()}' if c.isupper() else c for c in name[1:]
    )


def _to_konstant_case(name: str) -> str:
    """Convert to kTitleCase from a full name.

    Assumes local constants are already trimmed of their namespace.

    Examples:
    - 'nlb.buffham.constants.foo' -> 'nlb::buffham::constants::kFoo'
    - 'foo' -> 'kFoo'
    """
    if '.' in name:
        parts = name.split('.')
        return '::'.join(parts[:-2]) + '::k' + parts[-1].title().replace('_', '')

    return 'k' + name.title().replace('_', '')


def _generate_comment(comments: list[str], tabs: str) -> str:
    """Generate a comment block.

    Does not inlude a trailing newline.
    """
    definition = ''
    if len(comments) > 1:
        definition += f'{tabs}/*'
        for comment in comments:
            if comment:
                definition += f'\n{tabs}{comment.lstrip()}'
            else:
                definition += f'\n'
        definition += f'\n{tabs} */'
    elif comments:
        definition += f'{tabs}//{comments[0]}'
    return definition


def _get_namespaced_name(relative_name: str) -> str:
    """Get the namespaced name for a relative name."""
    split = relative_name.split('.')
    return '::'.join(split[:-2] + split[-1:])


def _cpp_type(field: parser.Field, primary_namespace: str) -> str:
    """Get the C++ type for the field."""
    if field.message:
        return _get_namespaced_name(field.message.get_relative_name(primary_namespace))

    if field.pri_type is parser.FieldType.LIST:
        assert field.sub_type is not None
        return f'std::vector<{TYPE_MAP[field.sub_type]}>'

    return TYPE_MAP[field.pri_type]


def generate_message(message: parser.Message, primary_namespace: str, hpp: bool) -> str:
    """Generate a struct definition from a Message."""
    definition = ''
    ns = '' if hpp else f'{message.name}::'
    tab = T if hpp else ''

    if hpp:
        if message.comments:
            definition += _generate_comment(message.comments, '') + '\n'

        definition += f'struct {message.name} {{'

        for field in message.fields:
            if field.comments:
                definition += '\n' + _generate_comment(field.comments, T)
            definition += f'\n{T}{_cpp_type(field, primary_namespace)} {field.name};'
            if field.inline_comment:
                definition += f'  // {field.inline_comment}'
        definition += '\n\n'

    # Add serializer method
    definition += (
        f'{tab}std::span<uint8_t> {ns}serialize(std::span<uint8_t> buffer) const'
    )

    if hpp:
        definition += ';\n\n'
    else:
        definition += ' {'
        offset = 0
        offset_str = ''
        for field in message.fields:
            if field.iterable:
                definition += (
                    f"\n{tab}{T}uint16_t {field.name}_size = {field.name}.size();"
                    f"\n{tab}{T}memcpy(buffer.data() + {offset}{offset_str}, &{field.name}_size, 2);"
                )
                offset += 2
                definition += f"\n{tab}{T}memcpy(buffer.data() + {offset}{offset_str}, {field.name}.data(), {field.name}_size * {field.size});"
                offset_str += f' + {field.name}_size * {field.size}'
            elif field.pri_type is parser.FieldType.MESSAGE:
                definition += f'\n{tab}{T}auto {field.name}_buffer = {field.name}.serialize(buffer.subspan({offset}{offset_str}));'
                offset_str += f' + {field.name}_buffer.size()'
            else:
                definition += f"\n{tab}{T}memcpy(buffer.data() + {offset}{offset_str}, &{field.name}, {field.size});"
                offset += field.size
        definition += f'\n{tab}{T}return buffer.subspan(0, {offset}{offset_str});\n'
        definition += f'{tab}}}\n\n'

    # Add deserializer method
    qualifiers = 'static ' if hpp else ''
    definition += f'{tab}{qualifiers}std::pair<{message.name}, std::span<const uint8_t> > {ns}deserialize(std::span<const uint8_t> buffer)'
    if hpp:
        definition += ';\n'
    else:
        definition += ' {'
        offset = 0
        offset_str = ''
        message_name = _to_snake_case(message.name)
        definition += f'\n{tab}{T}{message.name} {message_name};'
        for field in message.fields:
            if field.iterable:
                definition += (
                    f"\n{tab}{T}uint16_t {field.name}_size;"
                    f"\n{tab}{T}memcpy(&{field.name}_size, buffer.data() + {offset}{offset_str}, 2);"
                )
                offset += 2
                definition += (
                    f"\n{tab}{T}{message_name}.{field.name}.resize({field.name}_size);"
                )
                definition += f"\n{tab}{T}memcpy({message_name}.{field.name}.data(), buffer.data() + {offset}{offset_str}, {field.name}_size * {field.size});"
                offset_str += f' + {field.name}_size * {field.size}'
            elif field.pri_type is parser.FieldType.MESSAGE:
                definition += f'\n{tab}{T}auto {field.name}_buffer = buffer.subspan({offset}{offset_str});'
                definition += f'\n{tab}{T}std::tie({message_name}.{field.name}, {field.name}_buffer) = {_cpp_type(field, primary_namespace)}::deserialize({field.name}_buffer);'
                offset_str += f' + {field.name}_buffer.size()'
            else:
                definition += f"\n{tab}{T}memcpy(&{message_name}.{field.name}, buffer.data() + {offset}{offset_str}, {field.size});"
                offset += field.size

        definition += f'\n{tab}{T}return {{{message_name}, buffer.subspan(0, {offset}{offset_str})}};\n'
        definition += f'{tab}}}\n'

    if hpp:
        definition += '};\n'
    definition += '\n'

    return definition


def generate_project_class(
    name: str, transactions: list[parser.Transaction], primary_namespace: str
) -> str:
    """Generate a project class definition."""
    definition = (
        f'class {name} {{\n' f'{T[::2]}public:\n' f'{T}{name}();\n' f'{T}~{name}();\n\n'
    )

    # Add initialization method
    definition += f'{T}void initialize();\n\n'

    # Add register_handlers method
    definition += (
        f'{T}template <emb::network::serialize::SerializerLike S,\n'
        f'{T}{T}{T}{T}emb::network::transport::TransporterLike T, class... Projects>\n'
        f'{T}void register_handlers(emb::network::node::Node<S, T, Projects...> &node) {{\n'
    )
    for transaction in transactions:
        definition += (
            f'{T}{T}node.template register_handler<{_get_namespaced_name(transaction.receive.get_relative_name(primary_namespace))}, '
            f'{transaction.send.name}>({transaction.request_id}, '
            f'std::bind(&{name}::{transaction.name}, this, std::placeholders::_1));\n'
        )
    definition += f'{T}}}\n\n'

    # Add each transaction method
    for transaction in transactions:
        if transaction.comments:
            definition += _generate_comment(transaction.comments, T) + '\n'
        definition += f'{T}{_get_namespaced_name(transaction.send.get_relative_name(primary_namespace))} {transaction.name}(const {_get_namespaced_name(transaction.receive.get_relative_name(primary_namespace))} &{_to_snake_case(transaction.receive.name)});\n'

    # Add a pIMPL struct
    definition += f'{T[::2]}private:\n'
    definition += f'{T}struct {name}Impl;\n'
    definition += f'{T}{name}Impl *impl_;\n'

    definition += '};\n\n'

    return definition


def generate_publishes(publishes: list[parser.Publish]) -> str:
    """Generate a publish definition."""
    definition = ''

    if not publishes:
        return definition

    definition += 'enum PublishIds : uint8_t {\n'
    for publish in publishes:
        if publish.comments:
            definition += _generate_comment(publish.comments, T) + '\n'
        definition += f'{T}{publish.name.upper()} = {publish.request_id},\n'
    definition += '};\n\n'

    return definition


def generate_namespace(namespace: list[str]) -> str:
    """Generate a namespace definition."""
    definition = ''

    for part in namespace:
        definition += f'namespace {part} {{\n'

    definition += '\n'

    return definition


def generate_end_namespace(namespace: list[str]) -> str:
    """Generate an end namespace definition."""
    definition = ''

    for part in namespace:
        definition += f'}}  // namespace {part}\n'

    return definition


def generate_constant(constant: parser.Constant) -> str:
    """Generate a constant definition."""

    definition = ''

    if constant.comments:
        definition += _generate_comment(constant.comments, '') + '\n'
    references = {ref: _to_konstant_case(ref) for ref in constant.references}
    value = constant.value
    for ref, name in references.items():
        value = value.replace(f'{{{ref}}}', name)
    extra_type = ''
    if constant.type is parser.FieldType.STRING:
        value = f'"{value}"'
        # Gotta use string_view for constexpr strings
        extra_type = '_view'
    definition += f'constexpr {TYPE_MAP[constant.type]}{extra_type} {_to_konstant_case(constant.name)} = {value};'

    if constant.inline_comment:
        definition += f'  //{constant.inline_comment}'

    definition += '\n'

    return definition


def generate_cpp(
    ctx: parser.ParseContext, primary_namespace: str, outfile: pathlib.Path, hpp: bool
) -> None:
    bh = ctx.buffhams[primary_namespace]

    with outfile.open('w') as fp:
        if hpp:
            fp.write('#pragma once\n\n')
        else:
            fp.write(f'#include "{primary_namespace.replace(".", "/")}_bh.hpp"\n\n')

        if len(bh.messages):
            # Add includes
            # TODO: Trim down includes based on message types / move some
            # includes to the cc file
            fp.write(
                '#include <cinttypes>\n'
                '#include <cstring>\n'
                '#include <span>\n'
                '#include <string>\n'
                '#include <tuple>\n'
                '#include <vector>\n\n'
            )
        elif parser.FieldType.STRING in [constant.type for constant in bh.constants]:
            fp.write('#include <string>\n\n')

        if len(bh.transactions) and hpp:
            # Add includes
            fp.write(
                '#include "emb/network/node/node.hpp"\n'
                '#include "emb/network/serialize/serializer.hpp"\n'
                '#include "emb/network/transport/transporter.hpp"\n\n'
            )

        if len(ctx.buffhams) > 1:
            # Add includes for other namespaces
            for namespace in ctx.buffhams:
                if namespace != primary_namespace:
                    fp.write(f'#include "{namespace.replace(".", "/")}_bh.hpp"\n')
            fp.write('\n')

        fp.write(generate_namespace(bh.namespace.split('.')[:-1]))

        # Generate constant definitions
        if hpp:
            for constant in bh.constants:
                fp.write(generate_constant(constant))
            if bh.constants:
                fp.write('\n\n')

        # Generate message definitions
        for message in bh.messages:
            fp.write(generate_message(message, primary_namespace, hpp))

        # Generate transaction definitions
        if len(bh.transactions) and hpp:
            fp.write(
                generate_project_class(
                    bh.name.title(), bh.transactions, primary_namespace
                )
            )

        # Generate publish definitions
        if hpp:
            fp.write(generate_publishes(bh.publishes))

        fp.write(generate_end_namespace(bh.namespace.split('.')[:-1]))
