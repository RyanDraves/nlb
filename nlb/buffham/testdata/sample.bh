import nlb.buffham.testdata.other;

# This is a constant in the global scope
constant uint8_t my_constant = 4;

constant string constant_string = "Hello, world!";  # constants can have inline comments

# Constants may reference other constants with {brackets}
constant uint16_t composed_constant = 2 + {my_constant} + {nlb.buffham.testdata.other.other_constant};

# A message comment
message Ping {
    # Add some comments here
    uint8_t ping;
}

# This comment isn't associated with anything,
# and will be ignored by the parser

#
# A bunch of message comments,
# in a block-like pattern.
#
# All of these belong to `FlashPage`
#
message FlashPage {
    uint32_t address;
    # Another field comment
    list[uint8_t] data;  # What about some in-line comments for fields?
    # This comment belongs to `read_size`
    uint32_t read_size;
}

message LogMessage {  # In-line message comments are ignored
    string message;
}  # This in-line comment is also ignored

message NestedMessage {
    uint8_t flag;
    LogMessage message;
    list[int32_t] numbers;
    Ping pong;
    nlb.buffham.testdata.other.Ping other_pong;
}

transaction ping[nlb.buffham.testdata.other.Ping, LogMessage];
# Transaction comment
transaction flash_page[FlashPage, FlashPage];
transaction read_flash_page[FlashPage, FlashPage];  # In-line transaction comments are ignored
