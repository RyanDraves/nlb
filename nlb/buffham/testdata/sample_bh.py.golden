# @generated by Buffham
import dataclasses
import enum
import struct
from typing import Self, Type

from emb.network.serialize import bh_cobs
from emb.network.transport import transporter
from nlb.buffham import bh

from nlb.buffham.testdata import other_bh

# This is a constant in the global scope
MY_CONSTANT = 4
CONSTANT_STRING = "Hello, world!"  # constants can have inline comments
# Constants may reference other constants with {brackets}
COMPOSED_CONSTANT = 2 + MY_CONSTANT + other_bh.OTHER_CONSTANT


@dataclasses.dataclass
class Ping:
    """A message comment
    """

    # Add some comments here
    ping: int

    def serialize(self) -> bytes:
        buffer = bytes()
        buffer += struct.pack('<B', self.ping)
        return buffer

    @classmethod
    def deserialize(cls, buffer: bytes) -> tuple[Self, int]:
        ping = struct.unpack_from('<B', buffer, 0)[0]
        return cls(
            ping=ping,
        ), 1

@dataclasses.dataclass
class FlashPage:
    """
    A bunch of message comments,
    in a block-like pattern.

    All of these belong to `FlashPage`

    """

    address: int
    # Another field comment
    data: list[int]  # What about some in-line comments for fields?
    # This comment belongs to `read_size`
    read_size: int

    def serialize(self) -> bytes:
        buffer = bytes()
        buffer += struct.pack('<I', self.address)
        buffer += struct.pack('<H', len(self.data))
        buffer += struct.pack(f'<{len(self.data)}B', *self.data)
        buffer += struct.pack('<I', self.read_size)
        return buffer

    @classmethod
    def deserialize(cls, buffer: bytes) -> tuple[Self, int]:
        address = struct.unpack_from('<I', buffer, 0)[0]
        data_size = struct.unpack_from('<H', buffer, 4)[0]
        data = list(struct.unpack_from(f'<{data_size}B', buffer, 6))
        read_size = struct.unpack_from('<I', buffer, 6 + data_size * 1)[0]
        return cls(
            address=address,
            data=data,
            read_size=read_size,
        ), 10 + data_size * 1

@dataclasses.dataclass
class LogMessage:
    message: str

    def serialize(self) -> bytes:
        buffer = bytes()
        buffer += struct.pack('<H', len(self.message))
        buffer += self.message.encode()
        return buffer

    @classmethod
    def deserialize(cls, buffer: bytes) -> tuple[Self, int]:
        message_size = struct.unpack_from('<H', buffer, 0)[0]
        message = buffer[2:2 + message_size].decode()
        return cls(
            message=message,
        ), 2 + message_size * 1

@dataclasses.dataclass
class NestedMessage:
    flag: int
    message: LogMessage
    numbers: list[int]
    pong: Ping
    other_pong: other_bh.Pong

    def serialize(self) -> bytes:
        buffer = bytes()
        buffer += struct.pack('<B', self.flag)
        buffer += self.message.serialize()
        buffer += struct.pack('<H', len(self.numbers))
        buffer += struct.pack(f'<{len(self.numbers)}i', *self.numbers)
        buffer += self.pong.serialize()
        buffer += self.other_pong.serialize()
        return buffer

    @classmethod
    def deserialize(cls, buffer: bytes) -> tuple[Self, int]:
        flag = struct.unpack_from('<B', buffer, 0)[0]
        message, message_size = LogMessage.deserialize(buffer[1:])
        numbers_size = struct.unpack_from('<H', buffer, 1 + message_size)[0]
        numbers = list(struct.unpack_from(f'<{numbers_size}i', buffer, 3 + message_size))
        pong, pong_size = Ping.deserialize(buffer[3 + message_size + numbers_size * 4:])
        other_pong, other_pong_size = other_bh.Pong.deserialize(buffer[3 + message_size + numbers_size * 4 + pong_size:])
        return cls(
            flag=flag,
            message=message,
            numbers=numbers,
            pong=pong,
            other_pong=other_pong,
        ), 3 + message_size + numbers_size * 4 + pong_size + other_pong_size

REGISTRY: dict[int, Type[bh.BuffhamLike]] = {
    1: LogMessage,
    2: FlashPage,
    3: FlashPage,
    4: LogMessage,
}

class SampleSerializer(bh_cobs.BhCobs):
    def __init__(self, registry: bh_cobs.Registry | None = None):
        registry = (registry or {}) | REGISTRY | other_bh.REGISTRY
        super().__init__(registry)

class SampleNode[Transporter: transporter.TransporterLike](bh.BhNode[SampleSerializer, Transporter]):
    def __init__(self, serializer: SampleSerializer | None = None, transporter: Transporter | None = None):
        super().__init__(serializer or SampleSerializer(), transporter)

PING = bh.Transaction[other_bh.Pong, LogMessage](1)
# Transaction comment
FLASH_PAGE = bh.Transaction[FlashPage, FlashPage](2)
READ_FLASH_PAGE = bh.Transaction[FlashPage, FlashPage](3)

class PublishIds(enum.Enum):
    # Publish comment
    LOG_MESSAGE = 4
